server,*: untangle the Tracer from the Settings

andreimatei opened the pull request:
See individual commits.

andreimatei committed:
util/tracing: improve Tracer testing knobs

This patch makes the Tracer configurable with TestingKnobs at creation
time.
Also adds a new knob for forcing the creation of spans, to be used in
a next patch.

Release note: None

andreimatei committed:
testutils: refactor a LocalTestCluster method

Make NewDummyRaftTransport() take a Tracer explicitly, not through the
Settings. This is working towards untangling the settings from the
tracer.

Release note: None

andreimatei committed:
*: don't access Tracer through Settings

Our Tracer is unfortunately tangled with the Settings, for no good
reason. Creating a Settings also creates a Tracer, and the Settings also
exposes a Tracer field. Random code was using that field because they
needed a Tracer and they happened to have a Settings around.

This coupling is unfortunate because it's in the way of tests
controlling the creation of the Tracer. Also, it doesn't make any sense.
This patch moves the majority of users of Settings.Tracer to get a
Tracer in various other ways. The next patch will remove the coupling.

Release note: None

andreimatei committed:
sql: plumb tracer explicitly through EvalCtx

Some builtins needed the Tracer and was accessing it through the Settings.
I'm trying to untangle the Tracer from the Settings. This patch passes
the Tracer through the EvalCtx.

Release note: None

andreimatei committed:
server,*: untangle the Tracer from the Settings

Before this patch, out Settings had a Tracer field in them. Creating
Settings was also creating a Tracer. There was no good reason for this
coupling. The coupling was in the way of tests that want to control the
Tracer that a server will be constructed with.

This patch untangles the creation of a Tracer from the creation of
Settings. It also removes the last users for the Settings.Tracer field
(the majority had already gone away in the previous commit), and removes
the field.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71108)
<!-- Reviewable:end -->

andreimatei committed:
server,base: make the Tracer configurable in tests

This patch adds the Tracer to the TestServerArgs, allowing tests to
control the Tracer. This will allow tests to pass knobs to the Tracer
without having to go through SQL to modify cluster settings after the
cluster was already created.

Release note: None

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with nits

> Reviewed 5 of 5 files at r1, 8 of 8 files at r2, 15 of 15 files at r3, 13 of 13 files at r4, 4 of 4 files at r5, 6 of 6 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-:-MlPmusT0st8fdwRrajk:bfmeux1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 86 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm2ZM-y9h06Kc5LRJ:-MlPm2ZM-y9h06Kc5LRK:buikh04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L86)):*
> ```Go
> 	// Create a new DistSender and client.DB so that the Get below is guaranteed
> 	// to not hit in the range descriptor cache forcing a RangeLookup operation.
> 	ambient := log.AmbientContext{Tracer: s.Tracer().(*tracing.Tracer)}
> ```

.TracerTyped
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 968 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm3WzCsgF-91vGP4p:-MlPm3WzCsgF-91vGP4q:buikh04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L968)):*
> ```Go
> 				clock := hlc.NewClock(manual.UnixNano, time.Nanosecond)
> 				ds := kvcoord.NewDistSender(kvcoord.DistSenderConfig{
> 					AmbientCtx:         log.AmbientContext{Tracer: s.Tracer().(*tracing.Tracer)},
> ```

.TracerTyped
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 1177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm44t99rXw54L04Nw:-MlPm44t99rXw54L04Nx:buikh04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L1177)):*
> ```Go
> 	// we don't update the caches on the default dist sender-backed client.
> 	ds := kvcoord.NewDistSender(kvcoord.DistSenderConfig{
> 		AmbientCtx:         log.AmbientContext{Tracer: s.Tracer().(*tracing.Tracer)},
> ```

.TracerTyped
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go, line 68 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm6bZBz_i9f-eNWpJ:-MlPm6bZBz_i9f-eNWpK:buikh04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go#L68)):*
> ```Go
> 		defer s.Stopper().Stop(ctx)
> 
> 		tr := s.Tracer().(*tracing.Tracer)
> ```

.TracerTyped
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go, line 89 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm6zT1Zydqa6nkwUG:-MlPm6zT1Zydqa6nkwUH:buikh04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go#L89)):*
> ```Go
> 		}
> 
> 		tr := s.Tracer().(*tracing.Tracer)
> ```

.TracerTyped
___
*[pkg/server/admin_test.go, line 321 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm7gOCI_jLQ2vX35w:-MlPm7gPAijNzZELxov5:buikh04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/server/admin_test.go#L321)):*
> ```Go
> 	ts := s.(*TestServer)
> 
> 	ac := log.AmbientContext{Tracer: s.Tracer().(*tracing.Tracer)}
> ```

.TracerTyped
___
*[pkg/server/admin_test.go, line 651 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm8Df1SvMiSp3R8yk:-MlPm8Df1SvMiSp3R8yl:buikh04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/server/admin_test.go#L651)):*
> ```Go
> 			schemaName := "testschema"
> 
> 			ac := log.AmbientContext{Tracer: s.Tracer().(*tracing.Tracer)}
> ```

.TracerTyped
___
*[pkg/server/admin_test.go, line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm8esDE37yveFRKio:-MlPm8esDE37yveFRKip:buikh04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/server/admin_test.go#L797)):*
> ```Go
> 
> 	// Create database and table.
> 	ac := log.AmbientContext{Tracer: s.Tracer().(*tracing.Tracer)}
> ```

.TracerTyped
___
*[pkg/sql/flowinfra/cluster_test.go, line 93 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPlvgn4HAMUmvAdFdh:-MlPlvgn4HAMUmvAdFdi:buikh04) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/sql/flowinfra/cluster_test.go#L93)):*
> ```Go
> 
> 		// Start a span (useful to look at spans using Lightstep).
> 		sp := tc.Server(0).Tracer().(*tracing.Tracer).StartSpan("cluster test")
> ```

.TracerTyped
___
*[pkg/sql/sem/builtins/builtins.go, line 4335 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPlKUw2ywSatpE-cUE:-MlPlKUw2ywSatpE-cUF:b-1ekaiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/75451a2e3823d3a6243016566fa3e8ae3c1cb255/pkg/sql/sem/builtins/builtins.go#L4335)):*
> ```Go
> 				if ctx.Tracer == nil {
> 					// Let's not freak out if the ctx was not initialized fully.
> 					return nil, errors.Newf("Tracer not configured")
> ```

In a production server, should this not never happen?
Is this a case for an assertion error?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-:-MlQKx2F91cjjr6uMbxZ:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 86 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm2ZM-y9h06Kc5LRJ:-MlQ8cYnA-1wD6fvsPyV:br8teyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L86)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

.TracerTyped
</blockquote></details>

`s` is a `TestServerInterface`, not a `TestServer`, so it only has access to the untyped methods. But, for a bunch of tests in this file, `startNoSplitMergeServer(t)` can actually return a `TestServer`. I've done that here and also in some other tests files where I could.
I've also renamed `TestServerInterface.Tracer() -> TracerI()` and `TestServer.TracerTyped() -> Tracer()`. We already had precedent for the `FooI()` naming convention, and I disliked `TracerTyped()`.
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 968 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm3WzCsgF-91vGP4p:-MlQHPl782IBbGpGQY_A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L968)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

.TracerTyped
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/dist_sender_server_test.go, line 1177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm44t99rXw54L04Nw:-MlQHQwD0kRbxYX25gVJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go#L1177)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

.TracerTyped
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go, line 68 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm6bZBz_i9f-eNWpJ:-MlQJ4nhE4JqDOQf5Om9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go#L68)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

.TracerTyped
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go, line 89 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm6zT1Zydqa6nkwUG:-MlQJ52K2_ZGsBCWWPng:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go#L89)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

.TracerTyped
</blockquote></details>

Done.
___
*[pkg/server/admin_test.go, line 321 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm7gOCI_jLQ2vX35w:-MlQJIMuDZ_SsFZo53gk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/server/admin_test.go#L321)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

.TracerTyped
</blockquote></details>

Done.
___
*[pkg/server/admin_test.go, line 651 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm8Df1SvMiSp3R8yk:-MlQJNABAJRtD_aGv55S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/server/admin_test.go#L651)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

.TracerTyped
</blockquote></details>

Done.
___
*[pkg/server/admin_test.go, line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPm8esDE37yveFRKio:-MlQJLdT5Jia4biQQv87:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/server/admin_test.go#L797)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

.TracerTyped
</blockquote></details>

Done.
___
*[pkg/sql/flowinfra/cluster_test.go, line 93 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPlvgn4HAMUmvAdFdh:-MlQJSnKCljO-hvIyYd6:bhbpcjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56d0102b45d596002574c364d2142b411c5bf90/pkg/sql/flowinfra/cluster_test.go#L93)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

.TracerTyped
</blockquote></details>

done by introducing `TestCluster.ServerTyped()` besides the untyped `TestClusterInterface.Server()`.

I didn't boil the ocean to get other tests that can use `TestCluster` instead of `TestClusterInterface` to do so, though.
___
*[pkg/sql/sem/builtins/builtins.go, line 4335 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-MlPlKUw2ywSatpE-cUE:-MlQKr5g8BR-GDyH1XNI:b1nenpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/75451a2e3823d3a6243016566fa3e8ae3c1cb255/pkg/sql/sem/builtins/builtins.go#L4335)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

In a production server, should this not never happen?
Is this a case for an assertion error?
</blockquote></details>

yes; done


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71108#-:-MlQ_LM10we-bJ21eq9d:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz)



<!-- Sent from Reviewable.io -->
